src/dimension_slice.c

```diff
@@ -142,42 +222,56 @@ dimension_slice_copy(const DimensionSlice *original)
 	return new;
 }
 
-static int
-cmp_slices_by_dimension_id(const void *left, const void *right)
+/*
+ * Check if two dimensions slices overlap by doing collision detection in one
+ * dimension.
+ *
+ * Returns true if the slices collide, otherwise false.
+ */
+bool
+dimension_slices_collide(DimensionSlice *slice1, DimensionSlice *slice2)
 {
-	const DimensionSlice *left_slice = *((DimensionSlice **) left);
-	const DimensionSlice *right_slice = *((DimensionSlice **) right);
-
-	if (left_slice->fd.dimension_id == right_slice->fd.dimension_id)
-		return 0;
-	if (left_slice->fd.dimension_id < right_slice->fd.dimension_id)
-		return -1;
-	return 1;
-}
-
+	Assert(slice1->fd.dimension_id == slice2->fd.dimension_id);
 
-static void
-hypercube_slice_sort(Hypercube *hc)
-{
-	qsort(hc->slices, hc->num_slices, sizeof(DimensionSlice *), cmp_slices_by_dimension_id);
+	return (slice1->fd.range_start < slice2->fd.range_end &&
+			slice1->fd.range_end > slice2->fd.range_start);
 }
 
-Hypercube *
-hypercube_from_constraints(ChunkConstraint constraints[], int16 num_constraints)
+/*
+ * Cut a slice that collides with another slice. The coordinate is the point of
+ * insertion, and determines which end of the slice to cut.
+ *
+ * Case where we cut "after" the coordinate:
+ *
+ * |-X--------|
+ *		  |--------|
+ *
+ * Case where we cut "before" the coordinate:
+ *
+ *		 |------X--|
+ * |--------|
+ *
+ * Returns true if the slice was cut, otherwise false.
+ */
+bool
+dimension_slice_cut(DimensionSlice *to_cut, DimensionSlice *other, int64 coord)
 {
-	Hypercube  *hc = hypercube_alloc(num_constraints);
-	int			i;
-
-	for (i = 0; i < num_constraints; i++)
+	if (other->fd.range_end <= coord &&
```

Comment:
I think we should add Assert(to_cut->fd.dimension_id == other->fd.dimension_id);